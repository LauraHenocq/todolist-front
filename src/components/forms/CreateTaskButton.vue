<template>
  <v-container>
    <v-btn color="pink">Ajouter une tÃ¢che</v-btn>
  </v-container>
</template>