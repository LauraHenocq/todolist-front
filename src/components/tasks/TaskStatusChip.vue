<script setup lang="ts">
import { TaskStatus } from '@/models/task.entity';
import { computed } from 'vue';

type Props = {
  status: TaskStatus
}

const props = defineProps<Props>()

const getStatus = (): string => {
  switch (props.status) {
    case TaskStatus.TODO:
      return 'A faire'
    case TaskStatus.IN_PROGRESS:
      return 'En cours'
    case TaskStatus.DONE:
      return 'TerminÃ©e'
    default:
      return 'A faire'
  } 
}

const getStatusColor = computed(() => {
  switch (props.status) {
  case TaskStatus.TODO:
      return 'red'
    case TaskStatus.IN_PROGRESS:
      return 'secondary'
    case TaskStatus.DONE:
      return 'green'
    default:
      return 'grey'
  }
});
</script>

<template>
  <v-chip variant="flat" :color="getStatusColor">{{ getStatus(props.status) }}</v-chip>
</template>