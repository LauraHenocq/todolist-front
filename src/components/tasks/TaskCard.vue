<script setup lang="ts">
import { TaskEntity } from '@/models/task.entity';
import TaskStatusChip from '@/components/tasks/TaskStatusChip.vue';


type Props = {
  task: TaskEntity
}

const { task } = defineProps<Props>()

const emit = defineEmits(['delete-task', 'edit-task', 'go-to-task']);

const editTask = () => {
  emit('edit-task', task.id);
};

const deleteTask = () => {
  emit('delete-task', task.id);
};

const goToTask = () => {
  emit('go-to-task', task.id);
}

</script>
<template>
  <v-card>
    <div data-cy="task-card" class="task-card" @click="goToTask">
      <v-card-title>
        <span class="headline">{{ task.title }}</span>
      </v-card-title>
      <v-card-subtitle>
        <span>{{ task.description }}</span>
      </v-card-subtitle>
      <v-card-text>
        <TaskStatusChip :status="task.status" />
      </v-card-text>
    </div>
    <v-card-actions>
      <v-btn data-cy="task-card-edit-task-button" text @click="editTask(task)">Modifier</v-btn>
      <v-btn data-cy="task-card-delete-task-button" text @click="deleteTask(task.id)">Supprimer</v-btn>
    </v-card-actions>
  </v-card>
</template>
<style scoped>
.task-card {
  cursor: pointer;
}
</style>