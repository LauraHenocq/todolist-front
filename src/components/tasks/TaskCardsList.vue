<script setup lang="ts">
import { TaskEntity, TaskStatus } from '@/models/task.entity';
import TaskCard from '@/components/tasks/TaskCard.vue';

type Props = {
  tasks: TaskEntity[]
}

defineProps<Props>()

const emit = defineEmits(['task-clicked', 'delete-task']);

const emitTaskClicked = (taskId: string) => {
  emit('task-clicked', taskId)
};

const deleteTask = (taskId: string) => {
  emit('delete-task', taskId);
}

</script>

<template>
<v-container>
    <v-row>
      <v-col v-for="task in tasks" :key="task.id" cols="12" md="6">
        <TaskCard :task="task" @deleteTask="deleteTask"/>
      </v-col>
    </v-row>
  </v-container>
</template>
