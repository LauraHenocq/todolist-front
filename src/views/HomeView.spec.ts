import { mount, VueWrapper } from '@vue/test-utils';
import { describe, it, expect, vi, beforeEach } from 'vitest';
import Homeview from '@/views/HomeView.vue';
import CreateTaskButton from '@/components/forms/CreateTaskButton.vue';
import { createTestingPinia } from '@pinia/testing';
import { useTasksStore } from '@/stores/tasks.store';
import { createVuetify } from 'vuetify';
import { TaskEntity, TaskStatus } from '@/models/task.entity';
import RouteNames from '@/router/constants';
import { nextTick } from 'vue';

const vuetify = createVuetify();

const mockUseRouter = {
  push: vi.fn(),
};

vi.mock('vue-router', () => ({
  useRouter: () => mockUseRouter,
}));

describe('HomeView', () => {
  let wrapper: VueWrapper;
  const pinia = createTestingPinia();
  let tasksStore: ReturnType<typeof useTasksStore>;

  global.ResizeObserver = require('resize-observer-polyfill');

  beforeEach(() => {
    wrapper = mount(Homeview, {
      global: {
        plugins: [pinia, vuetify],
      },
    });
    tasksStore = useTasksStore();
    tasksStore.tasks = [new TaskEntity({ title: 'Task Title', description: 'Task Description', _status: TaskStatus.TODO, id: 'task-id' })];
  });

  describe('render', () => {
    it('should render', () => {
      expect(wrapper.html()).toMatchSnapshot();
    });

    it('should show loading spinner when data is loading', async () => {
      wrapper.vm.isLoading = true;
      await nextTick();
      expect(wrapper.html()).toMatchSnapshot();
    });
  });

  describe('create a task', () => {
    it('should open task creation dialog when the user clicks on the create task button', async () => {
      await wrapper.findComponent(CreateTaskButton).trigger('click');
      expect(wrapper.vm.isTaskCreationDialogOpen).toBe(true);
    });
    
    it('should create a task and show success alert when the user validates the creation', async () => {
      const newTaskInput = { title: 'Titre de la nouvelle tâche', description: 'Description de la nouvelle tâche' }
    
      await wrapper.vm.openTaskCreationDialog();
          
      await wrapper.vm.createTask(newTaskInput);
    
      expect(tasksStore.handleCreateTask).toHaveBeenCalled();
      expect(tasksStore.handleCreateTask).toHaveBeenCalledWith(newTaskInput);
      expect(wrapper.vm.alertInfo.title).toBe('Tâche créée');
      expect(wrapper.vm.isAlertVisible).toBe(true);
    });
  });

  describe('delete a task', () => {
    it('should open the task deletion dialog when the user cliks on the delete button', async () => {
      await wrapper.vm.openTaskDeletionDialog('task-id-1');
      expect(wrapper.vm.isTaskDeletionDialogOpen).toBe(true);
      expect(wrapper.vm.selectedTaskId).toBe('task-id-1');
    });
    
    it('should delete the task and show success alert when the user validates the deletion', async () => {
      await wrapper.vm.openTaskDeletionDialog('task-id-1');
      await wrapper.vm.deleteTask('task-id-1');
    
      expect(tasksStore.handleDeleteTask).toHaveBeenCalledWith('task-id-1');
      expect(wrapper.vm.alertInfo.title).toBe('Tâche supprimée');
      expect(wrapper.vm.isAlertVisible).toBe(true);
    });
  });

  describe('go to task information', () => {
    it('should navigate to task information when the user clicks on the task', async () => {
      wrapper.find('[data-cy="task-card"]').trigger('click');
          
      await nextTick();
    
      expect(mockUseRouter.push).toHaveBeenCalledWith({
        name: RouteNames.TASK_INFORMATION,
        params: {
          id: 'task-id',
        },
      });
    });
  });
});