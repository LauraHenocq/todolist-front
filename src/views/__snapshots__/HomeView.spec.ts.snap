// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`HomeView > render > should render 1`] = `
"<main>
  <div class="v-application v-theme--light v-layout v-layout--full-height v-locale--is-ltr app">
    <div class="v-application__wrap">
      <div>
        <!---->
        <!---->
        <!---->
        <!---->
        <div class="v-container v-locale--is-ltr"><button type="button" class="v-btn v-btn--elevated v-theme--light bg-pink v-btn--density-default v-btn--size-default v-btn--variant-elevated"><span class="v-btn__overlay"></span><span class="v-btn__underlay"></span>
            <!----><span class="v-btn__content" data-no-activator="">Ajouter une tâche</span>
            <!---->
            <!---->
          </button></div>
        <!--v-if-->
        <div class="v-card v-theme--light v-card--density-default v-card--variant-elevated mx-auto" style="width: 90%;">
          <!---->
          <div class="v-card__loader">
            <div class="v-progress-linear v-theme--light v-locale--is-ltr" style="top: 0px; height: 0px; --v-progress-linear-height: 2px;" role="progressbar" aria-hidden="true" aria-valuemin="0" aria-valuemax="100">
              <!---->
              <div class="v-progress-linear__background"></div>
              <div class="v-progress-linear__buffer" style="width: 0%;"></div>
              <transition-stub name="fade-transition" appear="false" persisted="false" css="true">
                <div class="v-progress-linear__indeterminate">
                  <div class="v-progress-linear__indeterminate long"></div>
                  <div class="v-progress-linear__indeterminate short"></div>
                </div>
              </transition-stub>
              <!---->
            </div>
          </div>
          <!---->
          <!---->
          <header class="v-toolbar v-toolbar--density-default bg-pink v-theme--light v-locale--is-ltr">
            <!---->
            <div class="v-toolbar__content" style="height: 64px;">
              <!---->
              <!---->
              <div class="v-toolbar-title">
                <div class="v-toolbar-title__placeholder">
                  <!---->Ma liste de tâches
                </div>
              </div>
              <div class="v-spacer"></div>
              <!---->
            </div>
            <transition-stub name="expand-transition" appear="false" persisted="false" css="true">
              <!---->
            </transition-stub>
          </header>
          <div class="v-container v-locale--is-ltr">
            <div class="v-row">
              <div class="v-col-md-6 v-col-12">
                <div data-v-1a4c953a="" class="v-card v-theme--light v-card--density-default v-card--variant-elevated">
                  <!---->
                  <div class="v-card__loader">
                    <div class="v-progress-linear v-theme--light v-locale--is-ltr" style="top: 0px; height: 0px; --v-progress-linear-height: 2px;" role="progressbar" aria-hidden="true" aria-valuemin="0" aria-valuemax="100">
                      <!---->
                      <div class="v-progress-linear__background"></div>
                      <div class="v-progress-linear__buffer" style="width: 0%;"></div>
                      <transition-stub name="fade-transition" appear="false" persisted="false" css="true">
                        <div class="v-progress-linear__indeterminate">
                          <div class="v-progress-linear__indeterminate long"></div>
                          <div class="v-progress-linear__indeterminate short"></div>
                        </div>
                      </transition-stub>
                      <!---->
                    </div>
                  </div>
                  <!---->
                  <!---->
                  <div data-v-1a4c953a="" data-cy="task-card" class="task-card">
                    <div data-v-1a4c953a="" class="v-card-title"><span data-v-1a4c953a="" class="headline">Task Title</span></div>
                    <div data-v-1a4c953a="" class="v-card-subtitle"><span data-v-1a4c953a="">Task Description</span></div>
                    <div data-v-1a4c953a="" class="v-card-text"><span data-v-1a4c953a="" class="v-chip v-theme--light bg-red v-chip--density-default v-chip--size-default v-chip--variant-flat" draggable="false"><!----><span class="v-chip__underlay"></span>
                      <!---->
                      <!---->
                      <div class="v-chip__content" data-no-activator="">A faire</div>
                      <!---->
                      <!----></span>
                    </div>
                  </div>
                  <div data-v-1a4c953a="" class="v-card-actions"><button data-v-1a4c953a="" type="button" class="v-btn v-btn--slim v-theme--light v-btn--density-default v-btn--size-default v-btn--variant-text" data-cy="task-card-edit-task-button"><span class="v-btn__overlay"></span><span class="v-btn__underlay"></span>
                      <!----><span class="v-btn__content" data-no-activator="">Modifier</span>
                      <!---->
                      <!---->
                    </button><button data-v-1a4c953a="" type="button" class="v-btn v-btn--slim v-theme--light v-btn--density-default v-btn--size-default v-btn--variant-text" data-cy="task-card-delete-task-button"><span class="v-btn__overlay"></span><span class="v-btn__underlay"></span>
                      <!----><span class="v-btn__content" data-no-activator="">Supprimer</span>
                      <!---->
                      <!---->
                    </button></div>
                  <!---->
                  <!----><span class="v-card__underlay"></span>
                </div>
              </div>
            </div>
          </div>
          <!---->
          <!----><span class="v-card__underlay"></span>
        </div>
      </div>
    </div>
  </div>
</main>"
`;

exports[`HomeView > render > should show loading spinner when data is loading 1`] = `
"<main>
  <div class="v-application v-theme--light v-layout v-layout--full-height v-locale--is-ltr app">
    <div class="v-application__wrap">
      <div>
        <div class="v-progress-circular v-progress-circular--indeterminate v-theme--light v-progress-circular--size-default text-primary" role="progressbar" aria-valuemin="0" aria-valuemax="100"><svg style="transform: rotate(calc(-90deg + 0deg));" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 45.714285714285715 45.714285714285715">
            <circle class="v-progress-circular__underlay" fill="transparent" cx="50%" cy="50%" r="20" stroke-width="5.714285714285714" stroke-dasharray="125.66370614359172" stroke-dashoffset="0"></circle>
            <circle class="v-progress-circular__overlay" fill="transparent" cx="50%" cy="50%" r="20" stroke-width="5.714285714285714" stroke-dasharray="125.66370614359172" stroke-dashoffset="125.66370614359172px"></circle>
          </svg>
          <!---->
        </div>
      </div>
    </div>
  </div>
</main>"
`;
